cmd/internal/find/types/entrySchema.go

```diff
@@ -0,0 +1,188 @@
+say back to B, then it is difficult, if not impossible, to coordinate all the
+updates in a single iteration. However if we let "result[N] == p(N) for all N
+such that either p(N) == true or N has no children" be our starting state, then
+we can iteratively evaluate result[V] for all other nodes V using the information
+provided by a previous iteration. The terminating condition is when an iteration
+does not update "result". This is possible if "result" contains all nodes N, or if
+there are some indeterminate nodes V. In the latter case, every node M in V is either
+part of a cycle, or there is more than one way to get to M from the root. In either
+case, we can say that result[M] == false for those nodes.
+
+NOTE: To see why result[M] == false for all M in V, we do a proof by contradiction.
+Assume result[M] == true for some M in V, and that M is part of cycle A-B-...-M...-C-A.
+Here, we see that one iteration would update all ancestors of M. A subsequent iteration
+would update M's descendants starting from C. Thus, all the nodes in the cycle are
+determinant, which contradicts our previous assumption of their indeterminacy (since
+those nodes are also part of V). The proof for the other case is similar.
+
+NOTE: The starting state formalizes the first condition for result[N] (and also notes
+that if N is a leaf, then result[N] == p(N)). Subsequent iterations check the second
+condition.
+*/
+func evaluateNodesToKeep(s *EntrySchema, p EntrySchemaPredicate) map[string]bool {
+	result := make(map[string]bool)
+	visited := make(map[string]bool)
+
+	// Set-up our starting state by evaluating p(N). Note that after this code,
+	// result[N] == p(N) for all nodes N such that p(N) == true or N has no
+	// children.
+	var applyPredicate func(s *EntrySchema)
+	applyPredicate = func(s *EntrySchema) {
+		if _, ok := visited[s.TypeID]; ok {
+			return
+		}
+		result[s.TypeID] = p(s)
+		if !result[s.TypeID] && len(s.Children) > 0 {
+			delete(result, s.TypeID)
+		}
+		visited[s.TypeID] = true
+		for _, child := range s.Children {
+			applyPredicate(child)
+		}
+	}
+	applyPredicate(s)
+
+	// Now we iteratively update "result".
+	var updateResult func(s *EntrySchema)
+	updateResult = func(s *EntrySchema) {
+		if _, ok := visited[s.TypeID]; ok {
+			return
+		}
+		visited[s.TypeID] = true
+		for _, child := range s.Children {
+			updateResult(child)
+			if result[child.TypeID] {
+				result[s.TypeID] = true
+			}
+		}
+	}
+	for {
+		visited = make(map[string]bool)
+		prevNumNodes := len(result)
+		updateResult(s)
+		newNodes := len(result) - prevNumNodes
+		if newNodes <= 0 {
+			// We've reached the terminating condition. Note that result[V] == false
+			// for an indeterminate node V since false is the 0-value for a "bool" type.
+			// Thus, we can just return here.
+			return result
+		}
+	}
+}
+
+// toMap returns a map of <typeID> => <childTypeIDs...> (i.e. its pre-serialized
```

Comment:
I think I might change the `fs/schema` endpoint's response to a similar format b/c it's more idiomatic with graph serialization. The original way was to make the `stree` command easy to implement. That'll be a separate PR though.